core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -837,51 +836,54 @@ public void run() {
             nextBackoffIntervalNanos = retryPolicy.initialBackoffNanos;
           }
 
-          RetryPlan retryPlan = makeRetryDecision(status, trailers);
-          if (retryPlan.shouldRetry) {
-            // The check state.winningSubstream == null, checking if is not already committed, is
-            // racy, but is still safe b/c the retry will also handle committed/cancellation
-            FutureCanceller scheduledRetryCopy;
+          if (!isHedging) {
+            RetryPlan retryPlan = makeRetryDecision(status, trailers);
+            if (retryPlan.shouldRetry) {
+              // The check state.winningSubstream == null, checking if is not already committed, is
+              // racy, but is still safe b/c the retry will also handle committed/cancellation
+              FutureCanceller scheduledRetryCopy;
+              synchronized (lock) {
+                scheduledRetry = scheduledRetryCopy = new FutureCanceller(lock);
+              }
+              scheduledRetryCopy.setFuture(
+                  scheduledExecutorService.schedule(
+                      new Runnable() {
+                        @Override
+                        public void run() {
+                          callExecutor.execute(
+                              new Runnable() {
+                                @Override
+                                public void run() {
+                                  // retry
+                                  Substream newSubstream =
+                                      createSubstream(substream.previousAttemptCount + 1);
+                                  drain(newSubstream);
+                                }
+                              });
+                        }
+                      },
+                      retryPlan.backoffNanos,
+                      TimeUnit.NANOSECONDS));
+              return;
+            }
+          } else {
+            HedgingPlan hedgingPlan = makeHedgingDecision(status, trailers);
+            if (hedgingPlan.isHedgeable) {
+              pushbackHedging(hedgingPlan.hedgingPushbackMillis);
+            }
             synchronized (lock) {
-              scheduledRetry = scheduledRetryCopy = new FutureCanceller(lock);
+              state = state.removeActiveHedge(substream);
+              // The invariant is whether or not #(Potential Hedge + active hedges) > 0.
+              // Once hasPotentialHedging(state) is false, it will always be false, and then
+              // #(state.activeHedges) will be decreasing. This guarantees that even there may be
+              // multiple concurrent hedges, one of the hedges will end up committed.
+              if (hedgingPlan.isHedgeable) {
+                if (hasPotentialHedging(state) || !state.activeHedges.isEmpty()) {
+                  return;
+                }
+                // else, no activeHedges, no new hedges possible, try to commit
+              } // else, fatal, try to commit
```

Comment:
No longer using "fatal" here, this is "else isHedgeable=false"

Reply:
Fixed.